example/viewer.ts

```diff
@@ -105,6 +123,311 @@ export class Viewer {
+      iClipPlane: clipPlane,
+      geometry: geometry,
+      color: '#0000FF',
+      mesh: plane,
+      position: new THREE.Vector3(0, 0, 0),
+      rotation: new THREE.Quaternion(),
+      scale: new THREE.Vector3(1, 1, 1),
+    });
+    this.scene.add(plane);
+  }
+
+  addClipCylinder(clipCylinder: IClipCylinder): void {
+    const geometry = new THREE.CylinderGeometry( 1, 1, 1, 32 );
+    const material = new THREE.MeshBasicMaterial({
+      color: '#ff00ff',
+      wireframe: true,
+    });
+    const cylinder = new THREE.Mesh( geometry, material );
+    this.clipCylinders.push({
+      iClipCylinder: clipCylinder,
+      geometry: geometry,
+      color: '#0000FF',
+      mesh: cylinder,
+      position: new THREE.Vector3(0, 0, 0),
+      rotation: new THREE.Quaternion(),
+      scale: new THREE.Vector3(1, 1, 1),
+    });
+    this.scene.add(cylinder);
+  }
+
+  updateClipMode() {
+    const pcos = this.pointClouds;
+    const numPcos = pcos.length;
+    if (numPcos > 0) {
+      const pointCloud = pcos[0];
+      if (pointCloud.material.clipMode === ClipMode.HIGHLIGHT_INSIDE) {
+        pointCloud.material.clipMode = ClipMode.CLIP_OUTSIDE;
+      } else if (pointCloud.material.clipMode === ClipMode.CLIP_OUTSIDE) {
+        pointCloud.material.clipMode = ClipMode.CLIP_INSIDE;
+      } else if (pointCloud.material.clipMode === ClipMode.CLIP_INSIDE) {
+        pointCloud.material.clipMode = ClipMode.HIGHLIGHT_INSIDE;
+      }
+    }
+  }
+
+  translateClipBox = (translation: Vector3, axis: string) => {
+    const currentPos = this.clipBoxes[0].position;
+    if (axis === 'x') {
+      this.clipBoxes[0].position = new Vector3(translation.x, currentPos.y, currentPos.z);
+    } else if (axis === 'y') {
+      this.clipBoxes[0].position = new Vector3(currentPos.x, translation.y, currentPos.z);
+    } else if (axis === 'z') {
+      this.clipBoxes[0].position = new Vector3(currentPos.x, currentPos.y, translation.z);
+    }
+    
+    this.clipBoxes[0].mesh.position.set(this.clipBoxes[0].position.x, this.clipBoxes[0].position.y, this.clipBoxes[0].position.z);
+
+    let iClipBox = this.clipBoxes[0].iClipBox;
+    const m = new Matrix4;
+    m.compose(
+      this.clipBoxes[0].position,
+      this.clipBoxes[0].rotation,
+      this.clipBoxes[0].scale,
+    );
+    iClipBox.matrix.getInverse(m);
+    this.pointClouds[0].material.setClipBoxes([iClipBox]);
+  }
+
+  scaleClipBox = (scale: number) => {
+    this.clipBoxes[0].scale = new Vector3(scale, scale, scale);
+    this.clipBoxes[0].mesh.scale.set(this.clipBoxes[0].scale.x, this.clipBoxes[0].scale.y, this.clipBoxes[0].scale.z);
+    let iClipBox = this.clipBoxes[0].iClipBox;
+    const m = new Matrix4;
+    m.compose(
+      this.clipBoxes[0].position,
+      this.clipBoxes[0].rotation,
+      this.clipBoxes[0].scale,
+    )
+    iClipBox.matrix.getInverse(m);
+    this.pointClouds[0].material.setClipBoxes([iClipBox]);
+  }
+
+  translateClipSphere = (translation: Vector3, axis: string) => {
```

Comment:
All these methods are making the viewer too big. I suggest removing `this.` references and moving them out into a separate file like `clip-helpers.ts` or `clip-transforms.ts`